### PR TITLE
 *: adjust scheduler interface

### PR DESCRIPTION
**Pull request by nolouch**
this PR do:
- let the scheduler control schedule interval by itself
- let the scheduler know limiter and decide it can be scheduled. our scheduler may product two type operator, leader or peer, the controller does not distinguish them.

### CODE REVIEW

File: server/schedule/scheduler.go [hunk 1]
```diff
@@ -58,16 +52,18 @@ type Scheduler interface {
...
 	GetType() string
-	GetInterval() time.Duration
+	GetMinInterval() time.Duration
+	GetNextInterval(interval time.Duration) time.Duration
 	GetResourceKind() core.ResourceKind
 	GetResourceLimit() uint64
 	Prepare(cluster Cluster) error
 	Cleanup(cluster Cluster)
 	Schedule(cluster Cluster) *Operator
+	IsAllowSchedule() bool
```
Comment (disksing) [inline server/schedule/scheduler.go]: I suppose we can remove `GetResourceKind()`, `GetResourceLimit()`, `IsAllowSchedule()` from interface, Schedulers can check limit by themselves.

Comment (nolouch) [reply server/schedule/scheduler.go]: the controller should know `IsAllowSchedule`. Directly return nil operator will change tick time.

File: server/schedule/scheduler.go [hunk 2]
```diff
@@ -81,10 +75,44 @@ func RegisterScheduler(name string, createFn CreateSchedulerFunc) {
...
+	l.Lock()
+	defer l.Unlock()
+	l.counts[op.ResourceKind()]++
+}
+
+// RemoveOperator decrease the count by kind
+func (l *Limiter) RemoveOperator(op *Operator) {
+	l.Lock()
+	defer l.Unlock()
+	l.counts[op.ResourceKind()]--
```
Comment (siddontang) [inline server/schedule/scheduler.go]: need consider < 0 here?

**Comment by disksing** [comment]
LGTM.

File: server/schedule/scheduler.go [hunk 3]
```diff
@@ -58,33 +52,70 @@ type Scheduler interface {
...
 	GetType() string
-	GetInterval() time.Duration
-	GetResourceKind() core.ResourceKind
-	GetResourceLimit() uint64
+	GetMinInterval() time.Duration
+	GetNextInterval(interval time.Duration) time.Duration
 	Prepare(cluster Cluster) error
 	Cleanup(cluster Cluster)
 	Schedule(cluster Cluster) *Operator
+	IsAllowSchedule() bool
```
Comment (siddontang) [inline server/schedule/scheduler.go]: IsScheduleAllowd

**Comment by Connor1996** [comment]
LGTM.

**Comment by disksing** [comment]
We are going to merge it after GA.

**Pull request merged**